### PR TITLE
add test for isl alive confirmation via round trip latency

### PR DESCRIPTION
**Pull request by andriidovhan**

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 1]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
         getIsl().srcSwitch.dpId | getIsl().srcPort | getIsl().dstSwitch.dpId | null      | "dst_port"
     }
 
+    @Tidy
+    @Tags(HARDWARE)
+    def "An ISL between noviflow switches doesn't go down when one switch is DEACTIVATED"(){
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: Remove 'noviflow' from everywhere.
1. It is not because the switch is noviflow, it is because we use round trip latency to determine the isl state
2. When looking for switch, look at the switch features in order to determine whether the switch supports round trip latency (it can be an OVS switch in the future). The switch feature name which is used by system now is NOVIFLOW_COPY_FIELD

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 2]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+        def relatedIsl = topology.getRelatedIsls(swToDeactivate)
+        def nonNoviflowIsl = relatedIsl.find { !it.dstSwitch.noviflow }
+
+        lockKeeper.knockoutSwitch(swToDeactivate)
+        def IsSwDeactivated = true
+        Wrappers.wait(WAIT_OFFSET) {
+            assert northbound.getSwitch(swToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        then: "All ISLs are still DISCOVERED because the system uses round trip latency for ISL alive confirmation"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: all  round trip latency isls

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 3]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
+        // make sure that ISLs are not in FAILED status after timeout
+        Wrappers.timedLoop(WAIT_OFFSET / 2) {
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: Why use timed loop here?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 4]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+            withPool {
+                (relatedIsl - nonNoviflowIsl).everyParallel { link ->
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: eachParallel. Or assert the whole loop rather than each isl

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 5]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+                    assert northbound.getLink(link).state == IslChangeType.DISCOVERED
+                }
+            }
+        }
+
+        and: "All ISLs where at least one switch is not Noviflow are FAILED"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: This is a bit misleading. You can have a situation when you pick a switch that do not have non-noviflow neighbors. In this case your test will state that it checked that `All ISLs where at least one switch is not Noviflow are FAILED`, but in fact it will skip this check

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 6]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+    def "An ISL between noviflow switches doesn't go down when one switch is DEACTIVATED"(){
+        given: "A noviflow switch"
+        def swToDeactivate = topology.activeSwitches.find { it.noviflow }
+
+        when: "Deactivate the src switch"
+        def relatedIsl = topology.getRelatedIsls(swToDeactivate)
+        def nonNoviflowIsl = relatedIsl.find { !it.dstSwitch.noviflow }
+
+        lockKeeper.knockoutSwitch(swToDeactivate)
+        def IsSwDeactivated = true
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: camelCase

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy [hunk 7]
```diff
@@ -651,6 +654,50 @@ class LinkSpec extends HealthCheckSpecification {
...
+        cleanup:
+        IsSwDeactivated && lockKeeper.reviveSwitch(swToDeactivate)
+        Wrappers.wait(discoveryInterval + WAIT_OFFSET) {
+            assert northbound.getSwitch(swToDeactivate.dpId).state == SwitchChangeType.ACTIVATED
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.DISCOVERED
+            }.size() == topology.islsForActiveSwitches.size() * 2
+        }
+    }
+
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: I'm surprised that no other tests required any refactoring. There should be tests that check that ISLs go to failed status when a switch fails. And they should become broken now

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 1 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 6 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 2 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 3 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: deleted

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 4 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 5 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: fixed in the `given` block

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy (hunk 7 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/LinkSpec.groovy]: > I'm surprised that no other tests required any refactoring.

I haven't investigated it yet

> There should be tests that check that ISLs go to failed status when a switch fails. And they should become 
broken now

agree

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        when: "Deactivate the dst switch"
+        lockKeeper.knockoutSwitch(dstSwToDeactivate)
+        def isDstSwDeactivated = true
+        Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+            assert northbound.getSwitch(dstSwToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        then: "The round trip latency ISL is FAILED (because round_trip_status is not available in DB for current ISL \
+on both switches)"
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: why you oversleep if you can wait?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        isSrcSwDeactivated && lockKeeper.reviveSwitch(srcSwToDeactivate)
+        isDstSwDeactivated && lockKeeper.reviveSwitch(dstSwToDeactivate)
+        Wrappers.wait(discoveryInterval + WAIT_OFFSET) {
+            assert northbound.getSwitch(srcSwToDeactivate.dpId).state == SwitchChangeType.ACTIVATED
+            assert northbound.getSwitch(dstSwToDeactivate.dpId).state == SwitchChangeType.ACTIVATED
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.DISCOVERED
+            }.size() == topology.islsForActiveSwitches.size() * 2
+        }
+    }
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: I'm not sure that I like the scenario. Isn't it better to simultaneously knockout 2 switches? The part that proves that ISL will remain DISCOVERED when 1 side is down is already covered in previous test, while simultaneous knockout may be an edge case (consult with Dmitriy)

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+                    it.dstSwitch.features.contains(SwitchFeature.NOVIFLOW_COPY_FIELD)
+                }
+                nonRoundTripIsls = topology.getRelatedIsls(sw).findAll {
+                    !it.dstSwitch.features.contains(SwitchFeature.NOVIFLOW_COPY_FIELD)
+                }
+                roundTripIsls && nonRoundTripIsls
+            }
+        } ?: assumeTrue("Wasn't able to find a switch with suitable links", false)
+
+        when: "Deactivate the switch"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: I don't think that we should call it 'Deactivate the switch' in these scenarios. Deactivating the switch may be understood as if the switch actually became down. In these scenarios it is very important that we actually _lose connection to switch_, but the switch itself remains operable.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        when: "Deactivate the src switch"
+        lockKeeper.knockoutSwitch(srcSwToDeactivate)
+        def isSrcSwDeactivated = true
+        // it takes more time to DEACTIVATE a switch via the 'knockoutSwitch' method on the stage env
+        Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+            assert northbound.getSwitch(srcSwToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        then: "The round trip ISL is still DISCOVERED (the system uses round trip latency for ISL alive confirmation)"
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: Is it mandatory to verify this again and again? It is actually a very time-consuming verification

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+
+        when: "Delete ROUND_TRIP_LATENCY_RULE_COOKIE on the dst switch"
+        northbound.deleteSwitchRules(dstSw.dpId, DeleteRulesAction.REMOVE_ROUND_TRIP_LATENCY)
+        def isRoundTripRuleDeleted = true
+        Wrappers.wait(RULES_DELETION_TIME) {
+            assert northbound.validateSwitch(dstSw.dpId).rules.missing.size() == 1
+        }
+
+        then: "The round trip latecny ISL is FAILED"
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: don't sleep. use wait.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+        }
+
+        when: "Install ROUND_TRIP_LATENCY_RULE_COOKIE on the dst switch"
+        northbound.installSwitchRules(dstSw.dpId, InstallRulesAction.INSTALL_ROUND_TRIP_LATENCY)
+        isRoundTripRuleDeleted = false
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            assert northbound.validateSwitch(dstSw.dpId).rules.missing.empty
+        }
+
+        then: "The round trip ISL is still FAILED"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: why? Also don't really understand what are you testing here if you will again remove this rule on next step

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+        northbound.getLink(roundTripIsl).state == IslChangeType.FAILED
+
+        when: "Delete ROUND_TRIP_LATENCY_RULE_COOKIE on the dst switch again"
+        northbound.deleteSwitchRules(dstSw.dpId, DeleteRulesAction.REMOVE_ROUND_TRIP_LATENCY)
+        isRoundTripRuleDeleted = true
+        Wrappers.wait(RULES_DELETION_TIME) {
+            assert northbound.validateSwitch(dstSw.dpId).rules.missing.size() == 1
+        }
+
+        and: "Activate the src switches"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: switch

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.DISCOVERED
+            }.size() == topology.islsForActiveSwitches.size() * 2
+        }
+        isSrcSwDeactivated = false
+
+        then: "Round trip isl is DISCOVERED again"
+        northbound.getLink(roundTripIsl).state == IslChangeType.DISCOVERED
+
+        and: "Round trip status is available again for the given ISL in both directions"
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: Why do you expect it to be active in both directions if you removed the RTL rule on L194 and never installed it back?

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+        isSrcSwDeactivated && lockKeeper.reviveSwitch(srcSwToDeactivate)
+        isRoundTripRuleDeleted && northbound.installSwitchRules(dstSw.dpId, InstallRulesAction.INSTALL_ROUND_TRIP_LATENCY)
+        Wrappers.wait(discoveryInterval + WAIT_OFFSET) {
+            assert northbound.getSwitch(srcSwToDeactivate.dpId).state == SwitchChangeType.ACTIVATED
+            assert northbound.getAllLinks().findAll {
+                it.state == IslChangeType.DISCOVERED
+            }.size() == topology.islsForActiveSwitches.size() * 2
+            assert northbound.validateSwitch(dstSw.dpId).rules.missing.empty
+        }
+    }
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: I don't understand the purpose of this test. The situation with 'missing RTL rule' is already like maximum strange. Ok, got 1 strange test for it. But this is a second test which does not test anything new over the ` "A round trip latency ISL goes down when the src switch is DEACTIVATED and round trip latency rule is removed \
on the dst switch"`. So I'm kinda confused by what is being tested here again

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+        and: "Deactivate the src switch"
+        lockKeeper.knockoutSwitch(srcSwToDeactivate)
+        def isSrcSwDeactivated = true
+        // it takes more time to DEACTIVATE a switch via the 'knockoutSwitch' method on the stage env
+        Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+            assert northbound.getSwitch(srcSwToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        then: "The round trip latency ISL is FAILED"
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: Review all the sleeps. Remove unrequired sleeps, reduce the amount of duplicated time-consuming verifications.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+
+        when: "Deactivate the src switch"
+        // it takes more time to DEACTIVATE a switch via the 'knockoutSwitch' method on the stage env
+        lockKeeper.knockoutSwitch(srcSwToDeactivate)
+        def isSrcSwDeactivated = true
+        Wrappers.wait(discoveryTimeout + WAIT_OFFSET) {
+            assert northbound.getSwitch(srcSwToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        then: "The round trip ISL is still DISCOVERED (the system uses round trip latency for ISL alive confirmation)"
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 8 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: my fault. fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 9 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: > Isn't it better to simultaneously knockout 2 switches?

agree, it would be better

test is refactored

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 11 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 12 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 13 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: deleted

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 17 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: fixed everywhere

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy (hunk 16 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: you are right, there is no much sense here(
test is deleted

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+
+        when: "Simulate connection lose between the switch and FL, the switch becomes DEACTIVATED and remains operable"
+        lockKeeper.knockoutSwitch(swToDeactivate)
+        def isSwDeactivated = true
+        Wrappers.wait(customWaitOffset) {
+            assert northbound.getSwitch(swToDeactivate.dpId).state == SwitchChangeType.DEACTIVATED
+        }
+
+        and: "Wait discoveryTimeout"
+        sleep((discoveryTimeout + WAIT_OFFSET / 2) * 1000 as long)
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: still don't understand the ` + WAIT_OFFSET / 2` thing. why not sleep for discoveryTimeout and then use _wait_ in order to wait for FAILED status.?

Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/links/RoundTripIslSpec.groovy]: I thought it would be better to do it in the same line and then do the verifications.
I am fine with your proposal. This line is refactored
